source/solvers/navier_stokes_assemblers.cc

```diff
@@ -1005,3 +1008,358 @@ BuoyancyAssembly<dim>::assemble_rhs(
+
+  // Loop and quadrature informations
+  const auto &       JxW_vec    = scratch_data.JxW;
+  const unsigned int n_q_points = scratch_data.n_q_points;
+  const unsigned int n_dofs     = scratch_data.n_dofs;
+  const double       h          = scratch_data.cell_size;
+  Tensor<2, dim>     identity;
+  for (unsigned int d = 0; d < dim; ++d)
+    {
+      identity[d][d] = 1;
+    }
+  std::vector<std::vector<Tensor<1, dim>>> prescribed_velocity_values;
+  prescribed_velocity_values =
+    std::vector<std::vector<Tensor<1, dim>>>(scratch_data.n_faces,
+                                             std::vector<Tensor<1, dim>>(
+                                               scratch_data.n_faces_q_points));
+
+  scratch_data.cell_size;
+  const double penalty_parameter =
+    1. / std::pow(scratch_data.cell_size * scratch_data.cell_size,
+                  double(dim - 1) / double(dim));
+  const FiniteElement<dim> &fe           = scratch_data.fe_face_values.get_fe();
+  auto &                    local_matrix = copy_data.local_matrix;
+  double                    beta         = boundary_conditions.beta;
+
+  for (unsigned int i_bc = 0; i_bc < this->boundary_conditions.size; ++i_bc)
+    {
+      if (this->boundary_conditions.type[i_bc] ==
+          BoundaryConditions::BoundaryType::function_weak)
+        {
+          for (unsigned int f = 0; f < scratch_data.n_faces; ++f)
+            {
+              if (scratch_data.is_boundary_face[f])
+                {
+                  if (scratch_data.boundary_face_id[f] ==
+                      this->boundary_conditions.id[i_bc])
+                    {
+                      for (unsigned int q = 0;
+                           q < scratch_data.n_faces_q_points;
+                           ++q)
+                        {
+                          const double JxW = scratch_data.face_JxW[f][q];
+                          for (const unsigned int i :
+                               scratch_data.fe_face_values.dof_indices())
+                            {
+                              const auto comp_i =
+                                fe.system_to_component_index(i).first;
+                              if (comp_i < dim)
+                                {
+                                  for (const unsigned int j :
+                                       scratch_data.fe_face_values
+                                         .dof_indices())
+                                    {
+                                      const auto comp_j =
+                                        fe.system_to_component_index(j).first;
+                                      if (comp_i == comp_j)
+                                        local_matrix(i, j) +=
+                                          penalty_parameter * beta *
+                                          scratch_data.face_phi_u[f][q][i] *
+                                          scratch_data.face_phi_u[f][q][j] *
+                                          JxW;
```

Comment:
Oh, you are only implementing the beta term.
There should be an additional term based on the gradient. Otherwise you are not coercive.

See this link:
https://www.dealii.org/current/doxygen/deal.II/step_70.html

Or the Nitsche-method paper
http://math.aalto.fi/~rstenber/Publications/Venice95.pdf

It's a quick fix that you could do in this PR :)

Reply:
done :)